plugins/datastreamer_plugin/utils/dynamic_types_utils.cpp

```diff
@@ -34,149 +35,260 @@ namespace eprosima {
+            parent_data,
+            members[members.size() - 1], // use last member Id (direct parent = parent_data)
+            actual_kind);    // use last kind (direct parent = parent_data)
     }
+}
+
+void get_introspection_string_data(
+        const TypeIntrospectionCollection& string_type_names,
+        const eprosima::fastrtps::types::DynamicData_ptr& data,
+        TypeIntrospectionStringStruct& string_data_result)
+{
+    DEBUG("Getting string data");
 
+    // It is used by index of the vector to avoid unnecessary lookups.
+    // Vectors must be sorted in the same order => creation order given in get_introspection_type_names
     for (int i = 0; i < string_type_names.size(); ++i)
     {
-        string_data[i].second = get_string_type_from_data(
+        const auto& member_type_info = string_type_names[i];
+
+        // Get reference to variables to avoid calling get twice
+        const std::vector<MemberId>& members =
+            std::get<std::vector<MemberId>>(member_type_info);
+        const std::vector<TypeKind>& kinds =
+            std::get<std::vector<TypeKind>>(member_type_info);
+
+        const TypeKind& actual_kind =
+            std::get<TypeKind>(member_type_info);
+
+        // Get Data parent that has the member we are looking for
+        const auto& parent_data = get_parent_data_of_member(
             data,
-            std::get<eprosima::fastrtps::types::MemberId>(string_type_names[i]),
-            std::get<eprosima::fastrtps::types::TypeKind>(string_type_names[i]));
+            members,
+            kinds);
+
+        string_data_result[i].second = get_string_type_from_data(
+            parent_data,
+            members[members.size() - 1], // use last member Id (direct parent = parent_data)
+            actual_kind);    // use last kind (direct parent = parent_data)
+    }
+}
+
+DynamicData_ptr get_parent_data_of_member(
+        const DynamicData_ptr& data,
+        const std::vector<MemberId>& members_tree,
+        const std::vector<TypeKind>& kind_tree,
+        unsigned int array_indexes /* = 0 */)
+{
+    // Get the member and kind of the current level
+    const MemberId& member_id = members_tree[array_indexes];
+    const TypeKind& kind = kind_tree[array_indexes];
+
+    if (array_indexes == members_tree.size() - 1)
+    {
+        // One-to-last value, so this one is the value to take, as it is the parent of the data
+        return data;
+    }
+    else
+    {
+        switch (kind)
+        {
+            // TODO
+        default:
+            // TODO add exception
+            WARNING("Error getting data");
+            return data;
+            break;
+        }
     }
 }
 
 double get_numeric_type_from_data(
-        eprosima::fastrtps::types::DynamicData_ptr data,
-        eprosima::fastrtps::types::MemberId member,
-        eprosima::fastrtps::types::TypeKind kind)
+        const DynamicData_ptr& data,
+        const MemberId& member,
+        const TypeKind& kind)
```

Comment:
NIT: fundamental types are normally passed by copy across interfaces instead of `const T&`, but it's ok to use `const T&` is `T` is expected to evolve to a more elaborate type

Reply:
I was doubting about it. Finally decided to do like this so code is less restricted of those data types (if they change in future, signature must not change)